README
.. image:: https://secure.travis-ci.org/tiddlyweb/tiddlyweb.png
    :target: http://travis-ci.org/tiddlyweb/tiddlyweb
    :alt: build status
.. image:: https://pypip.in/v/tiddlyweb/badge.png
    :target: https://crate.io/packages/tiddlyweb
    :alt: Package Version
.. image:: https://pypip.in/d/tiddlyweb/badge.png
    :target: https://crate.io/packages/tiddlyweb
    :alt: Package Downloads

* TiddlyWeb Home Site: http://tiddlyweb.com/
* TiddlyWeb Git Repo: http://github.com/tiddlyweb/tiddlyweb

For additional installation instructions see http://tiddlyweb.tiddlyspace.com/

Introduction
============

TiddlyWeb is an open source HTTP API for storing and accessing flexible
and composable microcontent. It is also a toolkit for tiddlers on the web.
Tiddlers are small chunks of optionally structured content
with their own URI. The concept comes from `TiddlyWiki <http://tiddlywiki.com>`_.
TiddlyWeb can provide the basis of a `server side for TiddlyWiki
<http://pypi.python.org/pypi/tiddlywebwiki>`_. TiddlyWeb provides:

* an abstract model (with default implementations) for:

  * authentication and authorization
  * entity storage 
  * entity and collection serialization

* a straightforward extension model via plugins
* an implementation of authenticated recipes and bags allowing
  access to dynamically created and filtered collections of tiddlers
* a clean and pragmatic HTTP API

The system is designed so that parts that are not optimal for a
particular installation can be easily improved or swapped out via
plugins.

While TiddlyWeb was initially designed as a TiddlyWiki server-side
it can also be used as a generic data store and platform. The platform
builds on concepts learned from TiddlyWiki, primarily the concept of
the tiddler: a small chunk of data used to build up a greater whole.

TiddlyWeb includes a command line tool called ``twanager``. Run ``twanager``
without arguments for a list of available commands.

Installation
============

The easiest way to install TiddlyWeb and all its dependencies is by
using `pip <http://pip.openplans.org/>`_ to install it from `PyPI
<http://pypi.python.org>`_:::

   pip install -U tiddlyweb

If you wish to use TiddlyWeb as a server-side for TiddlyWiki to generate
a multi-user TiddlyWiki system, you will also need `tiddlywebwiki
<http://pypi.python.org/pypi/tiddlywebwiki>`_. This too can be installed
via pip. Installing tiddlywebwiki will install tiddlyweb for you:::

   pip install -U tiddlywebwiki

Or you can install by hand, you will need the following requirements:

* Python 2.5, 2.6 or 2.7. (A Python 3 port exists but is incomplete,
  contact the authors if you are interested in the Python 3 version. Python
  2.4 may work but is neither tested nor recommended.)
* selector: http://lukearno.com/projects/selector/
* simplejson: http://undefined.org/python/#simplejson
* html5lib (for sanitizing input which may be rendered as HTML):
  http://code.google.com/p/html5lib/
* httpexceptor: http://pypi.python.org/pypi/httpexceptor

Building
=============

Tests are run using PyTest:

   make test

Coverage support with:

https://pypi.python.org/pypi/coverage
https://pypi.python.org/pypi/pytest-cov

Miscellaneous
=============

See http://pypi.python.org/pypi?%3Aaction=search&term=tiddlywebplugins
for plugins that TiddlyWeb can use with itself.

See http://tiddlyspace.com/ for a shared hosting service built on
TiddlyWeb.

If you have questions or contributions on making TiddlyWeb work
for you, please post to the TiddlyWeb group at google groups:

  http://groups.google.com/group/tiddlyweb

or contact the primary author, Chris Dent <cdent@peermore.com>.

Please report bugs and issues at the github issue tracker:

  https://github.com/tiddlyweb/tiddlyweb/issues

There are a few tools that make exploring TiddlyWeb a bit easier:

* HTTP API explorer: http://tiddlyweb.tiddlyspace.com/explorer
* Python API explorer: ``twanager interact`` (once in type ``locals().keys()``)
* HTTP API quick ref: http://tiddlyweb.tiddlyspace.com/HTTP%20API
* x-doc serialization: http://tiddlyweb.tiddlyspace.com/HTTP%20API.x-doc

License
=======
TiddlyWeb is released under the BSD License and is copyright
2008-2013 UnaMesa Association.


CHANGES
2.4.3 Wed Aug 18 15:47:24 UTC 2021

  * Fix handling of query strings in the face of ';' no
    longer being accepted in some stdlibs.

2.4.0 Sun Nov  9 16:24:42 UTC 2014

  * Add support for application/x-tiddler-dictionary as
    peudo binary type.

2.3.0 Mon May 12 15:30:54 UTC 2014

  * Switch default JSON Accept from application/json to
    application/vnd.tiddlyweb+json.

2.2.0 Mon May  5 16:34:26 UTC 2014

  * Allow Unicode usernames in cookie-based auth and extraction.

2.1.5 Sun Mar 30 15:30:23 UTC 2014

  * Guard against bad input in multipart-form processing.

2.1.3 Fri 14 Mar 2014 14:08:44 UTC

  * Allow UTF-8 fieldnames in text serialization.

2.1.2 Wed Mar  5 20:38:17 UTC 2014

  * Process multipart/form-data in POSTs.

2.1.1 Mon Mar  3 19:50:56 UTC 2014

  * Always send Etag and Last-Modified on tiddler GET.

2.1.0 Sat Feb 22 21:46:41 UTC 2014

  * Never accept modifier from incoming data on tiddler PUT.

2.0.2 Wed Feb  5 19:45:04 UTC 2014

  * Make policies compoarable.

2.0.1 Sat Jan 18 18:16:53 UTC 2014

  * Encoding tweaks to align with Python 3 expectations.
  * Make sure wsgiapp is included in the MANIFEST.

2.0.0 Sat Jan  4 14:13:12 UTC 2014

  * Add support for Python 3.3
  * Remove support for Python < 2.6
  * Rename apache.py to wsgiapp.py

1.4.18 Thu Oct 10 15:11:53 UTC 2013

  * Remove support for syslogd-based logging.
  * Finished pydoc for http://tiddlyweb.readthedocs.org/

1.4.17 Tue Sep 17 16:41:54 UTC 2013

  * Better handling of encoded % in URIs.
  * Start at improved pydoc.

1.4.16 Wed Aug  7 14:00:33 UTC 2013

  * deny storing a tiddler using reserved fields

1.4.12 Thu May 30 17:43:43 UTC 2013

  * More closely follow the httpbis spec for 304 responses.

1.4.11 Tue May 21 10:35:16 UTC 2013

  * Add 'wsgi_server' config entry, defaulting to wsgiref.
  * Remove support for building bundles.
  * Add 'collections.use_memory' config setting.

1.4.10 Fri May 17 16:40:43 UTC 2013

  * Peg html5lib version to continue Python 2.5 support.

1.4.9 Mon Apr 29 20:00:19 UTC 2013

  * Fix bug in get_serialize_type when missing or incomplete environ.

1.4.8 Tue Apr  9 11:37:10 UTC 2013

  * Ensure encode_name works correctly in Python 2.7.

1.4.7 Sun Mar  3 15:52:57 UTC 2013

  * Correct sys-logging on older Python.

1.4.6 Sun Feb 17 18:56:57 UTC 2013

  * Protect against unicode revision ids in etags.

1.4.5 Mon Feb 11 17:18:18 UTC 2013

  * Allow non integer revisions on tiddlers.

1.4.4 Tue Feb  5 18:16:40 UTC 2013

  * Add .bag and .recipe attributes to Tiddlers collection.

1.4.2 Tue Jan 29 14:43:24 UTC 2013

  * Add log_syslog to config to allow logging to syslogd.

1.4.1 Sat Jan 26 12:44:06 UTC 2013

  * Fix presentation of challengers to include HTML framing.

1.4.0 Mon Jan 21 17:59:17 UTC 2013

  * Remove HTMLPresenter in favor of more complete HTML serialization.
  * Add creator to text serialization.
  * Make text serialization more flexible (see tiddler_as).

1.3.6 Tue Jan  8 11:29:58 UTC 2013

  * Correct serialization bug when listing bags and recipes.

1.3.5 Fri Jan  4 19:42:13 UTC 2013

  * Refactored sendtiddlers.

1.3.4 Sun Dec 23 12:47:24 UTC 2012

  * Improve cookie form challenger's HTML.

1.3.3 Fri Dec 21 21:52:26 UTC 2012

  * Correct handling of _canonical_uri tiddlers.
  * Code cleanups (to make tidier modules).

1.3.2 Sat Dec 15 00:00:00 UTC 2012

  * Correct missing import.

1.3.1 Thu Dec 13 23:13:52 UTC 2012

  * More flexibility in `list_entities()`.

1.3.0 Thu Dec 13 18:03:54 UTC 2012

  * Extract httpexceptor to own module.
  * Add `twanager interact` interactive console.
  * More strict extension handing on collections.

1.2.67 Fri Nov  2 16:17:14 UTC 2012

  * Deal with bad If-Modified-Since headers. Thanks @fnd.

1.2.66 Tue Oct 30 14:19:06 UTC 2012

  * Handle unicode username error when creating ETags.

1.2.65 Mon Oct 15 17:29:16 UTC 2012

  * Ensure utf-8 charset on textual tiddlers.

1.2.64 Thu Jul 26 13:55:18 UTC 2012

  * Ensure server and host from command line are str. Thanks @pads

1.2.63 Tue Jun 26 17:54:00 UTC 2012

  * More strict unicode handling when initializing a new Tiddler.

1.2.62 Mon May 28 15:42:00 UTC 2012

  * Optional trailing slash on bags and recipes no longer allowed.
  * Extract reading wsgi.input to a util function.

1.2.61 Wed May  2 20:02:25 UTC 2012

  * Refactoring to make code tidier.
  * Rare race condition in sort filters fixed.

1.2.60 Fri Apr 20 17:46:45 UTC 2012

  * Better handing and reporting for some 400 errors.

1.2.59 Thu Apr  5 13:50:34 UTC 2012

  * Improved imports when using CherryPy.
  * More delicate and deliberate use of Cache-Control: no-transform.

1.2.58 Tue Mar 27 11:23:11 BST 2012

  * Handle TypeError when tiddler PUT.

1.2.57 Fri Mar  9 16:37:31 UTC 2012

  * Previous release didn't actually provide the described feature.

1.2.56 Fri Mar  9 16:06:33 UTC 2012

  * Add a fat=0 option on JSON representation of tiddler and tiddlers.
    This makes it possible to get a render member, without text.

1.2.55 Thu Feb 23 18:54:35 UTC 2012

  * Add Cache-Control: no-transform on all responses.

1.2.54 Thu Feb  9 15:28:00 UTC 2012

  * Allow UTF-8 in tiddlyweb_redirect on challengers.

1.2.53 Sat Dec 17 14:49:07 UTC 2011

  * Release with the real 1.2.52 changes included.

1.2.52 Thu Dec 15 20:16:34 UTC 2011

  * Improve performance when generating multiple etags in one request.
  * Improved test coverage.

1.2.51 Mon Oct 24 12:47:54 UTC 2011

  * Allow tiddlers to set their own Cache-Control max-age with
    _cache-max-age field

1.2.50 Thu Oct 13 11:00:06 UTC 2011

  * Remove a Python >2.4-ism.
  * Update README to be more correct.

1.2.49 Wed Sep 28 16:11:42 UTC 2011

  * Release mimeparse fixes. Last release didn't have changes merged.

1.2.48 Tue Sep 27 11:22:42 UTC 2011

  * mimeparse type preferences are ordered correction

1.2.47 Fri Sep 23 17:35:00 UTC 2011

  * Use mimeparse for processing Accept header on GET.

1.2.46 Fri Sep  9 15:36:09 UTC 2011


atom/atom.py
"""
Atom feeds for tiddlyweb.

The Atom code is borrowed from Django's django/utils/feedgenerator.py

  http://www.djangoproject.com/documentation/syndication_feeds/
  http://code.djangoproject.com/browser/django/trunk/django/utils/feedgenerator.py

Which appears to be licensed with

PYTHON SOFTWARE FOUNDATION LICENSE VERSION 2

Thanks to those guys for making a feed library that hides the 
nasty XML details.

"""

import time
import types
import urllib
import datetime
import re
import email.Utils

from xml.sax.saxutils import XMLGenerator


from tiddlyweb.serializations import SerializationInterface
from tiddlyweb.serializations.html import Serialization as HTMLSerialization
from tiddlyweb.wikklyhtml import wikitext_to_wikklyhtml

class Serialization(HTMLSerialization):

    def __init__(self, environ={}):
        self.environ = environ

    def list_recipes(self, recipes):
        pass

    def list_bags(self, bags):
        pass

    def recipe_as(self, recipe):
        pass

    def list_tiddlers(self, bag):
        """
        Turn the contents of a bag into an Atom Feed.
        """
        tiddlers = bag.list_tiddlers()
        recipe = tiddlers[0].recipe

        if recipe:
            feed = Atom1Feed(
                    title=u'Tiddlers in Recipe %s' % recipe,
                    link=u'%srecipes/%s/tiddlers' % (self._server_url(), iri_to_uri(recipe)),
                    language=u'en',
                    description=u'the tiddlers of recipe %s' % recipe
                    )
        else:
            feed = Atom1Feed(
                    title=u'Tiddlers in Bag %s' % bag.name,
                    link=u'%sbags/%s/tiddlers' % (self._server_url(), iri_to_uri(bag.name)),
                    language=u'en',
                    description=u'the tiddlers of bag %s' % bag.name
                    )
        for tiddler in tiddlers:
            self._add_tiddler_to_feed(feed, tiddler)

        # can we avoid sending utf-8 and let the wrapper handle it?
        return feed.writeString('utf-8')

    def tiddler_as(self, tiddler):
        if tiddler.recipe:
            link = u'%srecipes/%s/tiddlers/%s' % \
                    (self._server_url(), iri_to_uri(tiddler.recipe), iri_to_uri(tiddler.title))
        else:
            link = u'%sbags/%s/tiddlers/%s' % \
                    (self._server_url(), iri_to_uri(tiddler.bag), iri_to_uri(tiddler.title))
        feed = Atom1Feed(
                title=u'%s' % tiddler.title,
                link=link,
                language=u'en',
                description=u'tiddler %s' % tiddler.title
                )
        self._add_tiddler_to_feed(feed, tiddler)
        return feed.writeString('utf-8')

    def _add_tiddler_to_feed(self, feed, tiddler):
        if tiddler.recipe:
            tiddler_link = 'recipes/%s/tiddlers' % tiddler.recipe
            link = u'%srecipes/%s/tiddlers/%s' % \
                    (self._server_url(), iri_to_uri(tiddler.recipe), iri_to_uri(tiddler.title))
        else:
            tiddler_link = 'bags/%s/tiddlers' % tiddler.bag
            link = u'%sbags/%s/tiddlers/%s' % \
                    (self._server_url(), iri_to_uri(tiddler.bag), iri_to_uri(tiddler.title))

        if tiddler.type and tiddler.type != 'None':
            description = 'Binary Content'
        else:
            description = wikitext_to_wikklyhtml(self._server_url(), tiddler_link, tiddler.text)
        feed.add_item(title=tiddler.title,
                link=link,
                description=description,
                author_name=tiddler.modifier,
                pubdate=self._tiddler_datetime(tiddler.modified)
                )

    def _tiddler_datetime(self, date_string):
        return datetime.datetime(*(time.strptime(date_string, '%Y%m%d%H%M')[0:6]))

    def _server_url(self):
        try:
            server_host = self.environ['tiddlyweb.config']['server_host']
            port = str(server_host['port'])
            if port == '80' or port == '443':
                port = ''
            else:
                port = ':%s' % port
            host = '%s://%s%s/' % (server_host['scheme'], server_host['host'], port)
        except KeyError:
            host = '/'
        return host

"""
Atom feed generation from django.
"""

class SimplerXMLGenerator(XMLGenerator):
    def addQuickElement(self, name, contents=None, attrs=None):
        "Convenience method for adding an element with no children"
        if attrs is None: attrs = {}
        self.startElement(name, attrs)
        if contents is not None:
            self.characters(contents)
        self.endElement(name)

def force_unicode(s, encoding='utf-8', strings_only=False, errors='strict'):
    """
    Similar to smart_unicode, except that lazy instances are resolved to
    strings, rather than kept as lazy objects.

    If strings_only is True, don't convert (some) non-string-like objects.
    """
    if strings_only and isinstance(s, (types.NoneType, int, long, datetime.datetime, datetime.date, datetime.time, float)):
        return s
    if not isinstance(s, basestring,):
        if hasattr(s, '__unicode__'):
            s = unicode(s)
        else:
            s = unicode(str(s), encoding, errors)
    elif not isinstance(s, unicode):
        # Note: We use .decode() here, instead of unicode(s, encoding,
        # errors), so that if s is a SafeString, it ends up being a
        # SafeUnicode at the end.
        s = s.decode(encoding, errors)
    return s

def smart_str(s, encoding='utf-8', strings_only=False, errors='strict'):
    """
    Returns a bytestring version of 's', encoded as specified in 'encoding'.

    If strings_only is True, don't convert (some) non-string-like objects.
    """
    if strings_only and isinstance(s, (types.NoneType, int)):
        return s
    elif not isinstance(s, basestring):
        try:
            return str(s)
        except UnicodeEncodeError:
            return unicode(s).encode(encoding, errors)
    elif isinstance(s, unicode):
        return s.encode(encoding, errors)
    elif s and encoding != 'utf-8':
        return s.decode('utf-8', errors).encode(encoding, errors)
    else:
        return s

def iri_to_uri(iri):
    """
    Convert an Internationalized Resource Identifier (IRI) portion to a URI
    portion that is suitable for inclusion in a URL.

    This is the algorithm from section 3.1 of RFC 3987.  However, since we are
    assuming input is either UTF-8 or unicode already, we can simplify things a
    little from the full method.

    Returns an ASCII string containing the encoded result.
    """
    # The list of safe characters here is constructed from the printable ASCII
    # characters that are not explicitly excluded by the list at the end of
    # section 3.1 of RFC 3987.
    if iri is None:
        return iri
    return urllib.quote(smart_str(iri), safe='/#%[]=:;$&()+,!?*')


"""
Syndication feed generation library -- used for generating RSS, etc.

Sample usage:

>>> from django.utils import feedgenerator
>>> feed = feedgenerator.Rss201rev2Feed(
...     title=u"Poynter E-Media Tidbits",
...     link=u"http://www.poynter.org/column.asp?id=31",
...     description=u"A group weblog by the sharpest minds in online media/journalism/publishing.",
...     language=u"en",
... )
>>> feed.add_item(title="Hello", link=u"http://www.holovaty.com/test/", description="Testing.")
>>> fp = open('test.rss', 'w')
>>> feed.write(fp, 'utf-8')
>>> fp.close()
"""

def rfc2822_date(date):
    return email.Utils.formatdate(time.mktime(date.timetuple()))

def rfc3339_date(date):
    if date.tzinfo:
        return date.strftime('%Y-%m-%dT%H:%M:%S%z')
    else:
        return date.strftime('%Y-%m-%dT%H:%M:%SZ')

def get_tag_uri(url, date):
    "Creates a TagURI. See http://diveintomark.org/archives/2004/05/28/howto-atom-id"
    tag = re.sub('^http://', '', url)
    if date is not None:
        tag = re.sub('/', ',%s:/' % date.strftime('%Y-%m-%d'), tag, 1)
    tag = re.sub('#', '/', tag)
    return u'tag:' + tag

class SyndicationFeed(object):
    "Base class for all syndication feeds. Subclasses should provide write()"
    def __init__(self, title, link, description, language=None, author_email=None,
            author_name=None, author_link=None, subtitle=None, categories=None,
            feed_url=None, feed_copyright=None, feed_guid=None, ttl=None):
        to_unicode = lambda s: force_unicode(s, strings_only=True)
        if categories:
            categories = [force_unicode(c) for c in categories]
        self.feed = {
            'title': to_unicode(title),
            'link': iri_to_uri(link),
            'description': to_unicode(description),
            'language': to_unicode(language),
            'author_email': to_unicode(author_email),
            'author_name': to_unicode(author_name),
            'author_link': iri_to_uri(author_link),
            'subtitle': to_unicode(subtitle),
            'categories': categories or (),
            'feed_url': iri_to_uri(feed_url),
            'feed_copyright': to_unicode(feed_copyright),
            'id': feed_guid or link,
            'ttl': ttl,
        }
        self.items = []

    def add_item(self, title, link, description, author_email=None,
        author_name=None, author_link=None, pubdate=None, comments=None,
        unique_id=None, enclosure=None, categories=(), item_copyright=None, ttl=None):
        """
        Adds an item to the feed. All args are expected to be Python Unicode
        objects except pubdate, which is a datetime.datetime object, and
        enclosure, which is an instance of the Enclosure class.
        """
        to_unicode = lambda s: force_unicode(s, strings_only=True)
        if categories:
            categories = [to_unicode(c) for c in categories]
        self.items.append({
            'title': to_unicode(title),
            'link': iri_to_uri(link),
            'description': to_unicode(description),
            'author_email': to_unicode(author_email),
            'author_name': to_unicode(author_name),
            'author_link': iri_to_uri(author_link),
            'pubdate': pubdate,
            'comments': to_unicode(comments),
            'unique_id': to_unicode(unique_id),
            'enclosure': enclosure,
            'categories': categories or (),
            'item_copyright': to_unicode(item_copyright),
            'ttl': ttl,
        })

    def num_items(self):
        return len(self.items)

    def write(self, outfile, encoding):
        """
        Outputs the feed in the given encoding to outfile, which is a file-like
        object. Subclasses should override this.
        """
        raise NotImplementedError

    def writeString(self, encoding):
        """
        Returns the feed in the given encoding as a string.
        """
        from StringIO import StringIO
        s = StringIO()
        self.write(s, encoding)
        return s.getvalue()

    def latest_post_date(self):
        """
        Returns the latest item's pubdate. If none of them have a pubdate,
        this returns the current date/time.
        """
        updates = [i['pubdate'] for i in self.items if i['pubdate'] is not None]
        if len(updates) > 0:
            updates.sort()
            return updates[-1]
        else:
            return datetime.datetime.now()

class Enclosure(object):
    "Represents an RSS enclosure"
    def __init__(self, url, length, mime_type):
        "All args are expected to be Python Unicode objects"
        self.length, self.mime_type = length, mime_type
        self.url = iri_to_uri(url)

class Atom1Feed(SyndicationFeed):
    # Spec: http://atompub.org/2005/07/11/draft-ietf-atompub-format-10.html
    mime_type = 'application/atom+xml'
    ns = u"http://www.w3.org/2005/Atom"
    def write(self, outfile, encoding):
        handler = SimplerXMLGenerator(outfile, encoding)
        handler.startDocument()
        if self.feed['language'] is not None:
            handler.startElement(u"feed", {u"xmlns": self.ns, u"xml:lang": self.feed['language']})
        else:
            handler.startElement(u"feed", {u"xmlns": self.ns})
        handler.addQuickElement(u"title", self.feed['title'])
        handler.addQuickElement(u"link", "", {u"rel": u"alternate", u"href": self.feed['link']})
        if self.feed['feed_url'] is not None:
            handler.addQuickElement(u"link", "", {u"rel": u"self", u"href": self.feed['feed_url']})
        handler.addQuickElement(u"id", self.feed['id'])
        handler.addQuickElement(u"updated", rfc3339_date(self.latest_post_date()).decode('ascii'))
        if self.feed['author_name'] is not None:
            handler.startElement(u"author", {})
            handler.addQuickElement(u"name", self.feed['author_name'])
            if self.feed['author_email'] is not None:
                handler.addQuickElement(u"email", self.feed['author_email'])
            if self.feed['author_link'] is not None:
                handler.addQuickElement(u"uri", self.feed['author_link'])
            handler.endElement(u"author")
        if self.feed['subtitle'] is not None:
            handler.addQuickElement(u"subtitle", self.feed['subtitle'])
        for cat in self.feed['categories']:
            handler.addQuickElement(u"category", "", {u"term": cat})
        if self.feed['feed_copyright'] is not None:
            handler.addQuickElement(u"rights", self.feed['feed_copyright'])
        self.write_items(handler)
        handler.endElement(u"feed")

    def write_items(self, handler):
        for item in self.items:
            handler.startElement(u"entry", {})
            handler.addQuickElement(u"title", item['title'])
            handler.addQuickElement(u"link", u"", {u"href": item['link'], u"rel": u"alternate"})
            if item['pubdate'] is not None:
                handler.addQuickElement(u"updated", rfc3339_date(item['pubdate']).decode('ascii'))

            # Author information.
            if item['author_name'] is not None:
                handler.startElement(u"author", {})
                handler.addQuickElement(u"name", item['author_name'])
                if item['author_email'] is not None:
                    handler.addQuickElement(u"email", item['author_email'])
                if item['author_link'] is not None:
                    handler.addQuickElement(u"uri", item['author_link'])
                handler.endElement(u"author")

            # Unique ID.
            if item['unique_id'] is not None:
                unique_id = item['unique_id']
            else:
                unique_id = get_tag_uri(item['link'], item['pubdate'])
            handler.addQuickElement(u"id", unique_id)

            # Summary.
            if item['description'] is not None:
                handler.addQuickElement(u"summary", item['description'], {u"type": u"html"})

            # Enclosure.
            if item['enclosure'] is not None:
                handler.addQuickElement(u"link", '',
                    {u"rel": u"enclosure",
                     u"href": item['enclosure'].url,
                     u"length": item['enclosure'].length,
                     u"type": item['enclosure'].mime_type})

            # Categories.
            for cat in item['categories']:
                handler.addQuickElement(u"category", u"", {u"term": cat})

            # Rights.
            if item['item_copyright'] is not None:
                handler.addQuickElement(u"rights", item['item_copyright'])

            handler.endElement(u"entry")


